engine/runtime/src/main/java/org/enso/interpreter/node/callable/argument/ArgumentSorterNode.java

```diff
@@ -98,22 +98,37 @@ public MappedArguments execute(Function function, State state, Object[] argument
     if (argumentsExecutionMode.shouldExecute()) {
       executeArguments(arguments, state);
     }
-    Object[] mappedAppliedArguments = prepareArguments(function, arguments);
+    Object[] mappedAppliedArguments =
+        prepareArguments(
+            this.preApplicationSchema,
+            this.postApplicationSchema,
+            this.mapping,
+            function,
+            arguments);
     Object[] oversaturatedArguments = null;
     if (postApplicationSchema.hasOversaturatedArgs()) {
       oversaturatedArguments = generateOversaturatedArguments(function, arguments);
     }
     return new MappedArguments(mappedAppliedArguments, oversaturatedArguments);
   }
 
-  private Object[] prepareArguments(Function function, Object[] arguments) {
+  @ExplodeLoop
+  static Object[] prepareArguments(
+      FunctionSchema preSchema,
+      FunctionSchema postSchema,
+      ArgumentMapping map,
+      Function fn,
+      Object[] args) {
     Object[] mappedAppliedArguments;
-    if (preApplicationSchema.hasAnyPreApplied()) {
-      mappedAppliedArguments = function.clonePreAppliedArguments();
+    if (preSchema.hasAnyPreApplied()) {
+      mappedAppliedArguments = new Object[preSchema.getArgumentsCount()];
```

Comment:
Rather than cloning the `Function.getPreAppliedArguments()` array (which is _just some random data_), @kustosz suggests to use `preSchema.getArgumentsCount()` which is a _compilation constant_. As such the whole `mappedAppliedArguments` can be a compilation constant and the compiler can curry on optimizations.

Using `ValueProfile.profile(function)` works, but doesn't given any speedup (over statistical error) in the `CurriedFunctionBenchmarks`.